### DISCUSSION TITLE
How to create state watchers

### DISCUSSION THREAD

**Discussion by martin-dimi** [pmndrs/zustand#1557] [Q&A] [answered]
We have quite a few use cases in which certain states are inferred from others, for instance:

```
const useNodeStore = create()(
  (set, get) => ({
    nodes: someListOfNodesThatChangeALot,
    hasErrors: booleanInferedFromNodes, // infered through nodes.some(node => node.errors.length > 0)
  })
)  
```

We don't want our components to listen on `nodes` as that will trigger a lot of unnecessary rerenders. Instead, we want them only to rerender when `hasErrors` changes. There's this [addon library](https://github.com/Diablow/zustand-store-addons) that allows for "watchers" and "computedValues" which basically allow for `hasErrors` to get calculated every time `nodes` changes, hence, not causing unnecessary rerenders. 

I appreciate that zustand is not opinionated but are there any good standards for this? I'd rather not use extra libraries.

**Comment by dai-shi** [comment, marked as answer]
The primary solution for derived state is to use a selector function.
```js
const useNodeStore = create()(
  (set, get) => ({
    nodes: someListOfNodesThatChangeALot,
  })
)
const useHasErrors = () => useNodeStore(({ nodes }) => nodes.some(node => node.errors.length > 0))
```

**Reply by martin-dimi** [reply]
So simple :heart:
